File: apps/worker/src/ai-provider-response.ts
import { APICallError } from "ai";
import {
  type UpstreamProviderResponse as AiProviderResponse,
  upstreamProviderResponseFromBody,
  upstreamProviderResponseFromUnknown,
} from "./upstream-response";

type APICallErrorLike = {
  statusCode?: unknown;
  responseHeaders?: unknown;
  responseBody?: unknown;
  cause?: unknown;
  errors?: unknown;
};

// AI SDK 会把 provider 响应藏在 APICallError/cause/errors 不同层级；这里统一转成上游 helper 的内部形状再取 rawResponseText。
export function providerResponseFromFetchResponse(response: Response, body: string, bodyTruncated = false, secrets: readonly string[] = []): AiProviderResponse {
  return upstreamProviderResponseFromBody(response, body, bodyTruncated, secrets);
}

export function providerResponseFromError(error: unknown, secrets: readonly string[] = []): AiProviderResponse | null {
  const apiError = findAPICallError(error);
  if (!apiError) return null;
  return upstreamProviderResponseFromUnknown({
    status: apiError.statusCode,
    statusText: null,
    headers: apiError.responseHeaders,
    body: apiError.responseBody,
    bodyTruncated: false,
  }, secrets);
}

function findAPICallError(error: unknown, seen = new WeakSet<object>()): APICallErrorLike | null {
  if (!error || typeof error !== "object") return null;
  if (seen.has(error)) return null;
  seen.add(error);
  let fallback: APICallErrorLike | null = null;
  if (isAPICallError(error)) {
    // 优先返回带 responseBody 的层级；只有 status/header 的外层错误不应该遮住真正 provider body。
    if (hasProviderResponseBody(error)) return error;
    fallback = error;
  }
  if ("cause" in error) {
    const causeMatch = findAPICallError((error as { cause?: unknown }).cause, seen);
    if (causeMatch && hasProviderResponseBody(causeMatch)) return causeMatch;
    fallback ??= causeMatch;
  }
  if ("errors" in error) {
    const nestedErrors = (error as { errors?: unknown }).errors;
    if (Array.isArray(nestedErrors)) {
      for (const item of nestedErrors) {
        const itemMatch = findAPICallError(item, seen);
        if (itemMatch && hasProviderResponseBody(itemMatch)) return itemMatch;
        fallback ??= itemMatch;
      }
    }
  }
  return fallback;
}

function isAPICallError(error: unknown): error is APICallErrorLike {
  if (APICallError.isInstance(error)) return true;
  return Boolean(
    error
      && typeof error === "object"
      && (
        "responseBody" in error
          || "responseHeaders" in error
          || "statusCode" in error
      ),
  );
}

function hasProviderResponseBody(error: APICallErrorLike): boolean {
  return typeof error.responseBody === "string" && error.responseBody.length > 0;
}

File: apps/docker-server/cmd/renewlet/ai_provider_response.go
package main

// ai_provider_response.go 收敛 AI provider 原始响应回显契约。
//
// 该结构只随当前认证错误响应返回给前端，不写日志、不入库、不进导出；request headers/API key 仍不可回显。
import (
	"errors"
	"net/http"

	"github.com/zendev-sh/goai"
)

type aiProviderResponse = upstreamProviderResponse

func aiProviderResponseFromHTTPResponse(response *http.Response, body string, secrets ...string) *aiProviderResponse {
	if response == nil {
		return nil
	}
	return upstreamProviderResponseFromBody(response, redactAIRecognitionSecrets(body), false, secrets)
}

func aiProviderResponseFromError(err error) *aiProviderResponse {
	if err == nil {
		return nil
	}
	var apiErr *goai.APIError
	if errors.As(err, &apiErr) {
		status := apiErr.StatusCode
		return &aiProviderResponse{
			Status:        optionalAIProviderStatus(status),
			StatusText:    optionalUpstreamString(http.StatusText(status)),
			Headers:       upstreamHeaderMapToObject(apiErr.ResponseHeaders, nil),
			Body:          optionalUpstreamBody(redactAIRecognitionSecrets(apiErr.ResponseBody)),
			BodyTruncated: false,
		}
	}
	var overflowErr *goai.ContextOverflowError
	if errors.As(err, &overflowErr) {
		return &aiProviderResponse{
			Status:        nil,
			StatusText:    nil,
			Headers:       nil,
			Body:          optionalUpstreamBody(redactAIRecognitionSecrets(overflowErr.ResponseBody)),
			BodyTruncated: false,
		}
	}
	return nil
}

func optionalAIProviderStatus(status int) *int {
	if status < 100 || status > 599 {
		return nil
	}
	return &status
}
